### ISSUE TITLE
Singular matrix when calculating Delta indices

### ISSUE THREAD

**Issue by antonia-had** [SALib/SALib#240]
I'm submitting this as a new issue as #225 is closed.

So I updated to 1.3.4 as per @jdherman's suggestion and I still get the same singular matrix error (or, rather, the updated message). It’s doing it for both 1000 samples and 10000 samples of the same analysis done in different ways. There’s a lot of zeros in my output, so I thought that’s what’s causing it. 
I went to manually replicate the calculation where it breaks down. The script loops through every parameter and estimates the difference with the density function in the _calc_delta_ function below (where Y is the entire output and X is the samples for one of the parameters). The error is always produced when estimating the density function of Y at the specific X values (fyc = …)
In my case the Y was water shortage, and one parameter specifically is an on/off switch that [what it does doesn’t matter] but results in 0 shortages. gaussian_kde won’t work for an array of zeros (giving a singular matrix error) but the parameter indeed shifts the density function rather significantly and the difference should be added to d_hat. I get an error where I should actually be getting a very high index for that parameter. Does that make sense?
```
def calc_delta(Y, Ygrid, X, m):
    N = len(Y)
    fy = gaussian_kde(Y, bw_method='silverman')(Ygrid)
    xr = rankdata(X, method='ordinal')

    d_hat = 0
    for j in range(len(m) - 1):
        ix = np.where((xr > m[j]) & (xr <= m[j + 1]))[0]
        nm = len(ix)
        fyc = gaussian_kde(Y[ix], bw_method='silverman')(Ygrid)
        d_hat += (nm / (2 * N)) * np.trapz(np.abs(fy - fyc), Ygrid)

    return d_hat
```
Lastly, I don't know if it would be valuable to still report on the delta values of the other parameters even if one of them causes an error (for whatever reason). In my case all other parameters work fine, but this pesky one halts the whole thing.

**Comment by antonia-had** [comment]
Update:
This fixes the issue (checks if any value in Y[ix] is non-zero before estimating density function)
```
        if Y[ix].any():
            fyc = gaussian_kde(Y[ix], bw_method='silverman')(Ygrid)
            d_hat += (nm / (2 * N)) * np.trapz(np.abs(fy - fyc), Ygrid)
        else:
            d_hat += (nm / (2 * N)) * np.trapz(np.abs(fy), Ygrid)
```

**Comment by ConnectedSystems** [comment]
Hi @antonia-had 

Would you be able to share an example dataset? I'd like to add this situation as a test case.

Thanks

**Comment by antonia-had** [comment]
Here's a minimal working example:
Model output: [outputs.txt](https://github.com/SALib/SALib/files/3163195/outputs.txt)
Samples: [LHsamples.txt](https://github.com/SALib/SALib/files/3163196/LHsamples.txt)
```
import numpy as np
from SALib.analyze import delta

LHsamples = np.loadtxt('./LHsamples.txt')
output = np.loadtxt('./outputs.txt')
bounds = [[ 0.5 ,  1.5 ],[ 0.8 ,  1.  ],[ 0.5 ,  1.5 ],[ 0.5 ,  1.5 ],[ 0.  ,  1.  ],[ 0.  ,  1.  ],[-0.5 ,  1.  ],[ 0.98,  1.02],[ 0.75,  1.25],[ 0.98,  1.02],[ 0.75,  1.25],[-0.3 ,  0.3 ],[-0.3 ,  0.3 ],[ 0.  , 60.  ]]
problem = {
    'num_vars': len(LHsamples[0,:]),
    'names': list(np.arange(14)),
    'bounds': bounds
}
delta.analyze(problem, LHsamples, output, print_to_console=True, num_resamples=2)
```
This produces an error with the original code, but not with my "fix".
The parameter I mentioned in the OP is in the 6th column of the sample file.

**Comment by ConnectedSystems** [comment]
Hi @antonia-had thanks for the files.

I confirm the issue exists and is resolved by the proposed fix, which is implemented [here](https://github.com/ConnectedSystems/SALib/blob/0c836d9763f9a67b840eb911c48eeb029da171e5/src/SALib/analyze/delta.py#L105)

@jdherman would you be able to review what's going on here? If it's acceptable, I will set up a test specific for this issue

**Comment by jdherman** [comment]
This is great, thanks @ConnectedSystems . Go ahead and include a test when you get the chance. I see you also opened another PR (thanks!) -- should I wait to merge that one along with this?

**Comment by ConnectedSystems** [comment]
Hi @jdherman - unfortunately I leave for a conference in China tomorrow. I was hoping to finish this off before I leave but it looks like this will have to wait until I can find some spare time.

On the other PR, you can merge that one - it's just addressing a deprecated import from scipy.
Incidentally, I submitted it to the `develop` branch, let me know if you want it submitted to `master` instead.

**Comment by jdherman** [comment]
Sounds good, no rush. Have fun at the conference. Thanks!

**Comment by ConnectedSystems** [comment]
@jdherman please close this issue as well as it is resolved by PR #251
